cmd/shell.go

```diff
@@ -305,9 +314,107 @@ func ReadConfigFile(filename string, out interface{}) error {
 	}
 	defer file.Close()
 
-	decoder := json.NewDecoder(file)
-	decoder.DisallowUnknownFields()
-	return decoder.Decode(out)
+	return decodeJSONStrict(file, out)
+}
+
+// ReadAndValidateConfigFile takes a file path as an argument and attempts to
+// unmarshal the content of the file into a struct containing a configuration of
+// a boulder component. Any config keys in the JSON file which do not correspond
+// to expected keys in the config struct will result in errors. It also
+// validates the config using the struct tags defined in the config struct.
+func ReadAndValidateConfigFile(name, filename string) error {
+	file, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+	if name == "boulder-observer" {
+		// Only the boulder-observer uses YAML config files.
+		return ValidateYAMLConfigByName(name, file)
+	}
+	return ValidateJSONConfigByName(name, file)
+}
+
+// ValidateJSONConfigByName takes a config name and an io.Reader containing a JSON
```

Comment:
"config name" is not a defined term - probably this should be "command name"?

Actually now that I mention it, there's no requirement that a config type is registered using the same name as a command, though AFAICT that is always the case in practice.

Reply:
Fixed the comments here.